### PR TITLE
openvr crash on exit fix

### PR DESCRIPTION
**Pull request by hyperlogic**
Increased thread-safety for display plugin getHeadPose().

Before this fix, a script could call into HMD.getHUDLookAtPosition2D() while the app was shutting down, which in turn would call getHeadPose() on the currently active display plugin.  This call could cause a crash within the openvr plugin, because the SDK was either previously shutdown, or in the process of shutting down on the main thread.

This fixes this by splitting the previous DisplayPlugin::getHeadPose(int) into two parts:

* updateHeadPose(int) which is only called once a frame and only by the main thread.
* getHeadPose() which is thread-safe and will return a cached copy of the hmd pose sampled by the last updateHeadPose.

### CODE REVIEW

**Comment by hyperlogic** [comment]
https://app.asana.com/0/45663370065758/101905459350870

**Comment by hifi-gustavo** [comment]
Build console output for commit: 960ffd9c9e41054046d5238a03b2028e5fecb40c available at: https://gist.github.com/7c1bdd9cbf791d1a6f51<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.dmg

**Comment by ZappoMan** [comment]
```
/Users/jenkins/Jenkins/workspace/beta_pr-master/label/osx/plugins/oculusLegacy/src/OculusLegacyDisplayPlugin.h:29:23: warning: 'OculusLegacyDisplayPlugin::getHeadPose' hides overloaded virtual function [-Woverloaded-virtual]
In file included from /Users/jenkins/Jenkins/workspace/beta_pr-master/label/osx/plugins/oculusLegacy/src/OculusProvider.cpp:16:
/Users/jenkins/Jenkins/workspace/beta_pr-master/label/osx/libraries/plugins/src/plugins/DisplayPlugin.h:128:23: note: hidden overloaded virtual function 'DisplayPlugin::getHeadPose' declared here: different number of parameters (0 vs 1)
    virtual glm::mat4 getHeadPose() const {
                      ^
1 warning and 1 error generated.

** BUILD FAILED **
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4321c5a62b873c82246f00a73bc194d6e0e56638 available at: https://gist.github.com/7c1bdd9cbf791d1a6f51<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e1d47e702813f4f68a5588464d23b7beb000ae48 available at: https://gist.github.com/7c1bdd9cbf791d1a6f51<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.dmg

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -1418,6 +1418,12 @@ void Application::paintGL() {
     // FIXME not needed anymore?
     _offscreenContext->makeCurrent();
 
+    // Tell the plugin what pose we're using to render.  In this case we're just using the
```
Comment (jherico) [inline interface/src/Application.cpp]: comment is no longer accurate

File: plugins/oculus/src/OculusBaseDisplayPlugin.cpp [hunk 2]
```diff
@@ -15,14 +15,15 @@ void OculusBaseDisplayPlugin::resetSensors() {
...
-    if (frameIndex > lastFrameSeen) {
-        lastFrameSeen = frameIndex;
-    }
-    return toGlm(trackingState.HeadPose.ThePose);
+    auto trackingState = ovr_GetTrackingState(_session, displayTime, true);
+    mat4 headPose = toGlm(trackingState.HeadPose.ThePose);
+    _headPoseCache.set(headPose);
+}
+
+glm::mat4 OculusBaseDisplayPlugin::getHeadPose() const {
```
Comment (jherico) [inline plugins/oculus/src/OculusBaseDisplayPlugin.cpp]: should be moved to `HmdDisplayPlugin`

File: plugins/oculus/src/OculusBaseDisplayPlugin.h [hunk 3]
```diff
@@ -36,4 +38,5 @@ class OculusBaseDisplayPlugin : public HmdDisplayPlugin {
     ovrHmdDesc _hmdDesc;
     ovrLayerEyeFov _sceneLayer;
     ovrViewScaleDesc _viewScaleDesc;
+    ThreadSafeValueCache<glm::mat4> _headPoseCache { glm::mat4() };
```
Comment (jherico) [inline plugins/oculus/src/OculusBaseDisplayPlugin.h]: should be moved to `HmdDisplayPlugin`

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5b2a050df7161cc9d647839e9fa8c53245dff725 available at: https://gist.github.com/7c1bdd9cbf791d1a6f51<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7395.dmg

**Comment by cozza13** [comment]
I tested this on the Vive machine. With prod, on shutdown I crashed. When I went to this PR I did not crash on shutdown.

**Pull request merged**